sympy/series/sequences.py

```diff
@@ -711,6 +712,137 @@ def coeff_mul(self, coeff):
 
 
+class RecursiveSeq(SeqBase):
+
+    """A finite degree recursive sequence.
+
+    That is, a sequence a(n) that depends on a fixed, finite number of its
+    previous values. The general form is
+
+        a(n) = f(a(n - 1), a(n - 2), ..., a(n - d))
+
+    for some fixed, positive integer d, where f is some function defined by a
+    SymPy expression.
+
+    Examples
+    ========
+
+    >>> from sympy import Function, symbols
+    >>> from sympy.series.sequences import RecursiveSeq
+
+    >>> y = Function("y")
+    >>> n = symbols("n")
+    >>> fib = RecursiveSeq((0, 1), y(n), y(n - 1) + y(n - 2))
+
+    For value at a particular point
+
+    >>> fib.coeff(3)
+    2
+
+    supports slicing
+
+    >>> fib[:6]
+    [0, 1, 1, 2, 3, 5]
+
+    inspect recurrence
+
+    >>> fib.recurrence
+    y(n - 2) + y(n - 1)
+
+    automatically determine degree
+
+    >>> fib.degree
+    2
+
+    See Also
+    ========
+
+    sympy.series.sequences.SeqFormula
+
+    """
+
+    def __new__(cls, initial, y, recurrence, start=0, stop=oo):
+        n = y.args[0]
+        y = y.func
+        k = Wild("k", exclude=(n,))
+        degree = -oo
+
+        # Find all applications of y in the recurrence and check that:
+        #   1. The function y is only being used with a single argument; and
+        #   2. All arguments are n + k for constant negative integers k.
+
+        prev_ys = recurrence.find(y)
+        for prev_y in prev_ys:
+            if len(prev_y.args) != 1:
+                raise TypeError("Recurrence should be in a single variable")
+
+            shift = prev_y.args[0].match(n + k)[k]
+            if not (shift.is_constant() and shift.is_integer and shift < 0):
+                raise TypeError("Recurrence should have constant,"
+                                " negative, integer shifts"
+                                " (found {})".format(prev_y))
+
+            if -shift > degree:
+                degree = -shift
+
+        if len(initial) != degree:
+            raise ValueError("Number of initial terms must equal degree")
+
+        degree = S(degree)
+        start = sympify(start)
+        stop = sympify(stop)
+
+        initial = Tuple(*(sympify(x) for x in initial))
+
+        seq = Basic.__new__(cls, initial, y(n), recurrence, start, stop)
+
+        seq.cache = {y(start + k): init for k, init in enumerate(initial)}
+        seq._start = start
+        seq._stop = stop
+        seq.degree = degree
+        seq.y = y
+        seq.n = n
+        seq.recurrence = recurrence
+
+        return seq
+
+    @property
+    def start(self):
+        """The starting point of the sequence. This point is included"""
+        return self._start
+
+    @property
+    def stop(self):
+        """The ending point of the sequence. This point is included"""
+        return self._stop
+
+    @property
+    def interval(self):
+        """Interval on which sequence is defined."""
+        return (self._start, self._stop)
+
+    def _eval_coeff(self, index):
```

Comment:
Could this method be simplified using `@recurrence_memo`?

Reply:
It likely could; I'll check that decorator out more closely and see if it applies.

Reply:
I implemented a version with this, but it's a little clunky.

- To pass `@recurrence_memo` the initial values, `_eval_coeff` needs to be an inline function in `__new__`.

- Since `@recurrence_memo` indexes from 0, it's necessary to wrap the decorated function to shift indices down, and then shift them back up in the function itself.

If there was a version of `@recurrence_memo` that used a dictionary, or handled arbitrary starting indices, then this would be much cleaner.